src/filter.rs

```diff
@@ -0,0 +1,297 @@
+//! Interfaces for adding custom filter to libgit2
+
+use libc::{c_int, c_char, c_void, size_t};
+use std::{mem, slice, ptr};
+use {raw, call, util, Error};
+use std::ffi::{CString, CStr};
+use std::path::PathBuf;
+use raw::{git_filter, git_writestream};
+
+
+#[repr(C)]
+struct Filter {
+    filter: git_filter,
+    instance: Box<StreamFilter>,
+}
+
+#[repr(C)]
+struct Stream {
+    parent: git_writestream,
+    handler: Box<WriteStream>,
+}
+
+/// Wrapper of the raw git_writestream object, representing the input
+/// of next filter of the chain
+pub struct RawWriteStream {
+    stream: *mut git_writestream,
+}
+
+/// A trait representing operations to handle filter data stream.
+///
+/// These callbacks are typically called by libgit2.
+pub trait WriteStream : Send + 'static {
+    /// New stream data, Implementations should process the buffer data
+    /// and then write the output to the next stream if needed.
+    fn write(&mut self, buf: &[u8])
+        -> Result<(), Error>;
+
+    /// Close stream. Implementations should do finalizing work and then
+    /// close the next stream.
+    fn close(&mut self) -> Result<(), Error>;
+}
+
+/// Information to construct a new stream filter.
+pub struct FilterSource {
+    /// Path to the source file to be filter.
+    pub path: PathBuf,
+    /// A sumgde filter if true, otherwise a clean filter.
+    pub to_worktree: bool,
```

Comment:
Also, to mirror the design of the rest of this library, perhaps this should just contain the raw value? It looks like it may also need a lifetime parameter to ensure it's not sent to other threads.